src/main/scala/zio/prelude/Idempotent.scala

```diff
@@ -0,0 +1,498 @@
+package zio.prelude
+
+import zio.prelude.coherent.EqualIdempotent
+import zio.test.TestResult
+import zio.test.laws.{ Lawful, Laws }
+
+/**
+ * The `Idempotent` type class describes a binary operator for a type `A` that
+ * is both associative and produces the same value when combining two identical values.
+ * This means that `a <> a` is equal to `a` for all values `a`.
+ * Example of idempotent operations is union of sets, but not addition of integers.
+ *
+ * Idempotent operators are useful because combining  the values with an idempotent
+ * operation results in the same value regardless of the number of values
+ * are combined, allowing us to optimize out unnecessary combinations of the same values.
+ */
+trait Idempotent[A] extends Associative[A] { self =>
+
+  protected def combineNormal(l: => A, r: => A): A = combine(l, r)
```

Comment:
If it is really idempotent do we need `combineNormal` and `combineIdempotent`? Can't we just know that the `combine` operation is idempotent if the type extends `Idempotent`?

Reply:
`combineNormal` is a detail of the internal implementation that allows `def idempotent(...` to work properly.

We need `combineIdempotent` to have a method that omits unnecessary combining (but to do that, it needs `Equal[A]`).